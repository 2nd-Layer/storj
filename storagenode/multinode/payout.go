// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

package multinode

import (
	"context"
	"time"

	"go.uber.org/zap"

	"storj.io/common/rpc/rpcstatus"
	"storj.io/storj/private/multinodepb"
	"storj.io/storj/storagenode/apikeys"
	"storj.io/storj/storagenode/payouts"
	"storj.io/storj/storagenode/payouts/estimatedpayouts"
)

var _ multinodepb.DRPCPayoutServer = (*PayoutEndpoint)(nil)

// PayoutEndpoint implements multinode payouts endpoint.
//
// architecture: Endpoint
type PayoutEndpoint struct {
	multinodepb.DRPCPayoutUnimplementedServer

	log              *zap.Logger
	apiKeys          *apikeys.Service
	estimatedPayouts *estimatedpayouts.Service
	db               payouts.DB
}

// NewPayoutEndpoint creates new multinode payouts endpoint.
func NewPayoutEndpoint(log *zap.Logger, apiKeys *apikeys.Service, estimatedPayouts *estimatedpayouts.Service, db payouts.DB) *PayoutEndpoint {
	return &PayoutEndpoint{
		log:              log,
		apiKeys:          apiKeys,
		estimatedPayouts: estimatedPayouts,
		db:               db,
	}
}

// Earned returns total earned amount.
func (payout *PayoutEndpoint) Earned(ctx context.Context, req *multinodepb.EarnedRequest) (_ *multinodepb.EarnedResponse, err error) {
	defer mon.Task()(&ctx)(&err)

	if err = authenticate(ctx, payout.apiKeys, req.GetHeader()); err != nil {
		return nil, rpcstatus.Wrap(rpcstatus.Unauthenticated, err)
	}

	earned, err := payout.db.GetTotalEarned(ctx)
	if err != nil {
		return nil, rpcstatus.Wrap(rpcstatus.Internal, err)
	}

	return &multinodepb.EarnedResponse{
		Total: earned,
	}, nil
}

// EarnedPerSatellite returns total earned amount per satellite.
func (payout *PayoutEndpoint) EarnedPerSatellite(ctx context.Context, req *multinodepb.EarnedPerSatelliteRequest) (_ *multinodepb.EarnedPerSatelliteResponse, err error) {
	defer mon.Task()(&ctx)(&err)

	if err = authenticate(ctx, payout.apiKeys, req.GetHeader()); err != nil {
		return nil, rpcstatus.Wrap(rpcstatus.Unauthenticated, err)
	}

	var resp multinodepb.EarnedPerSatelliteResponse
	satelliteIDs, err := payout.db.GetPayingSatellitesIDs(ctx)
	if err != nil {
		return nil, rpcstatus.Wrap(rpcstatus.Internal, err)
	}

	for i := 0; i < len(satelliteIDs); i++ {
		earned, err := payout.db.GetEarnedAtSatellite(ctx, satelliteIDs[i])
		if err != nil {
			return nil, rpcstatus.Wrap(rpcstatus.Internal, err)
		}

		resp.EarnedSatellite = append(resp.EarnedSatellite, &multinodepb.EarnedSatellite{
			Total:       earned,
			SatelliteId: satelliteIDs[i],
		})
	}

	return &resp, nil
}

// EstimatedPayoutTotal returns estimated earnings for current month from all satellites.
func (payout *PayoutEndpoint) EstimatedPayoutTotal(ctx context.Context, req *multinodepb.EstimatedPayoutTotalRequest) (_ *multinodepb.EstimatedPayoutTotalResponse, err error) {
	defer mon.Task()(&ctx)(&err)

	if err = authenticate(ctx, payout.apiKeys, req.GetHeader()); err != nil {
		return nil, rpcstatus.Wrap(rpcstatus.Unauthenticated, err)
	}

	estimated, err := payout.estimatedPayouts.GetAllSatellitesEstimatedPayout(ctx, time.Now())
	if err != nil {
		return &multinodepb.EstimatedPayoutTotalResponse{}, rpcstatus.Wrap(rpcstatus.Internal, err)
	}

	return &multinodepb.EstimatedPayoutTotalResponse{EstimatedEarnings: estimated.CurrentMonthExpectations}, nil
}

// EstimatedPayoutSatellite returns estimated earnings for current month from specific satellite.
func (payout *PayoutEndpoint) EstimatedPayoutSatellite(ctx context.Context, req *multinodepb.EstimatedPayoutSatelliteRequest) (_ *multinodepb.EstimatedPayoutSatelliteResponse, err error) {
	defer mon.Task()(&ctx)(&err)

	if err = authenticate(ctx, payout.apiKeys, req.GetHeader()); err != nil {
		return nil, rpcstatus.Wrap(rpcstatus.Unauthenticated, err)
	}

	estimated, err := payout.estimatedPayouts.GetSatelliteEstimatedPayout(ctx, req.SatelliteId, time.Now())
	if err != nil {
		return &multinodepb.EstimatedPayoutSatelliteResponse{}, rpcstatus.Wrap(rpcstatus.Internal, err)
	}

	return &multinodepb.EstimatedPayoutSatelliteResponse{EstimatedEarnings: estimated.CurrentMonthExpectations}, nil
}
