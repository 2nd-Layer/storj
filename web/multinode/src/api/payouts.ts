// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import { APIClient } from '@/api/index';
import { PayoutsSummary } from '@/payouts';

/**
 * client for nodes controller of MND api.
 */
export class PayoutsClient extends APIClient {
    private readonly ROOT_PATH: string = '/api/v0/payouts';

    /**
     * handles fetch of payouts summary information.
     *
     * @param satelliteId - satellite id.
     * @param period - selected period.
     *
     * @throws {@link BadRequestError}
     * This exception is thrown if the input is not a valid.
     *
     * @throws {@link UnauthorizedError}
     * Thrown if the auth cookie is missing or invalid.
     *
     * @throws {@link InternalError}
     * Thrown if something goes wrong on server side.
     */
    public async summary(satelliteId: string | null, period: string | null): Promise<PayoutsSummary> {
        const path = `${this.ROOT_PATH}/summary?satelliteId=${satelliteId}&period=${period}`;

        const response = await this.http.get(path);

        if (!response.ok) {
            await this.handleError(response);
        }

        return await response.json();
    }
}
