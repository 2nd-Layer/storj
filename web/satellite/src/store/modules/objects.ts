// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import { GatewayCredentials } from '@/types/accessGrants';

export const OBJECTS_ACTIONS = {
    CLEAR: 'clearObjects',
    SET_GATEWAY_CREDENTIALS: 'setGatewayCredentials',
    SET_ACCESS_GRANT: 'setAccessGrant',
};

export const OBJECTS_MUTATIONS = {
    SET_GATEWAY_CREDENTIALS: 'setGatewayCredentials',
    SET_ACCESS_GRANT: 'setAccessGrant',
    CLEAR: 'clearObjects',
};

const {
    CLEAR,
    SET_ACCESS_GRANT,
    SET_GATEWAY_CREDENTIALS,
} = OBJECTS_MUTATIONS;

export class ObjectsState {
    public accessGrant: string = '';
    public gatewayCredentials: GatewayCredentials = new GatewayCredentials();
}

/**
 * Creates objects module with all dependencies.
 */
export function makeObjectsModule(): StoreModule<ObjectsState> {
    return {
        state: new ObjectsState(),
        mutations: {
            [SET_ACCESS_GRANT](state: ObjectsState, accessGrant: string) {
                state.accessGrant = accessGrant;
            },
            [SET_GATEWAY_CREDENTIALS](state: ObjectsState, credentials: GatewayCredentials) {
                state.gatewayCredentials = credentials;
            },
            [CLEAR](state: ObjectsState) {
                state.accessGrant = '';
                state.gatewayCredentials = new GatewayCredentials();
            },
        },
        actions: {
            setAccessGrant: function({commit}: any, accessGrant: string): void {
                commit(SET_ACCESS_GRANT, accessGrant);
            },
            setGatewayCredentials: function({commit}: any, credentials: GatewayCredentials): void {
                commit(SET_GATEWAY_CREDENTIALS, credentials);
            },
            clearObjects: function ({commit}: any): void {
                commit(CLEAR);
            },
        },
    };
}
