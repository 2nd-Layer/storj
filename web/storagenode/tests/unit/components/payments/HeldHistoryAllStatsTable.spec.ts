// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import HeldHistoryAllStatsTable from '@/app/components/payments/HeldHistoryAllStatsTable.vue';

import { makePayoutModule, PAYOUT_MUTATIONS } from '@/app/store/modules/payout';
import { HeldHistory, HeldHistoryAllStatItem } from '@/app/types/payout';
import { PayoutHttpApi } from '@/storagenode/api/payout';
import { createLocalVue, shallowMount } from '@vue/test-utils';

const localVue = createLocalVue();
localVue.use(Vuex);

localVue.filter('centsToDollars', (cents: number): string => {
    return `$${(cents / 100).toFixed(2)}`;
});

const payoutApi = new PayoutHttpApi();
const payoutModule = makePayoutModule(payoutApi);

const store = new Vuex.Store({ modules: { payoutModule }});

describe('HeldHistoryAllStatsTable', (): void => {
    it('renders correctly with actual values', async (): Promise<void> => {
        const wrapper = shallowMount(HeldHistoryAllStatsTable, {
            store,
            localVue,
        });

        const testJoinAt = new Date(Date.UTC(2020, 0, 30));

        await store.commit(PAYOUT_MUTATIONS.SET_HELD_HISTORY, new HeldHistory(
            [],
            [
                new HeldHistoryAllStatItem('1', 'name1', 1, 50000, 20000, testJoinAt),
                new HeldHistoryAllStatItem('2', 'name2', 5, 40000, 30000, testJoinAt),
            ],
        ));

        expect(wrapper).toMatchSnapshot();
    });
});
