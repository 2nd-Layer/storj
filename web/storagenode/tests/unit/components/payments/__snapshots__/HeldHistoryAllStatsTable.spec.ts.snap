// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeldHistoryAllStatsTable renders correctly with actual values 1`] = `
<div>
  <div class="held-history-table-container--large">
    <div class="held-history-table-container--large__labels-area">
      <div class="column justify-start column-1">
        <p class="held-history-table-container--large__labels-area__text">Satellite</p>
      </div>
      <div class="column justify-end column-2">
        <p class="held-history-table-container--large__labels-area__text">First Contact</p>
      </div>
      <div class="column justify-end column-3">
        <p class="held-history-table-container--large__labels-area__text">Held Total</p>
      </div>
      <div class="column justify-end column-4">
        <p class="held-history-table-container--large__labels-area__text">Held Returned</p>
      </div>
    </div>
    <div class="held-history-table-container--large__info-area">
      <div class="justify-start column-1">
        <p class="held-history-table-container--large__info-area__text">name1</p>
        <p class="held-history-table-container--large__info-area__months">1 month</p>
      </div>
      <div class="column justify-end column-2">
        <p class="held-history-table-container--large__info-area__text">2020-01-30</p>
      </div>
      <div class="column justify-end column-3">
        <p class="held-history-table-container--large__info-area__text">$500.00</p>
      </div>
      <div class="column justify-end column-4">
        <p class="held-history-table-container--large__info-area__text">$200.00</p>
      </div>
    </div>
    <div class="held-history-table-container--large__info-area">
      <div class="justify-start column-1">
        <p class="held-history-table-container--large__info-area__text">name2</p>
        <p class="held-history-table-container--large__info-area__months">5 month</p>
      </div>
      <div class="column justify-end column-2">
        <p class="held-history-table-container--large__info-area__text">2020-01-30</p>
      </div>
      <div class="column justify-end column-3">
        <p class="held-history-table-container--large__info-area__text">$400.00</p>
      </div>
      <div class="column justify-end column-4">
        <p class="held-history-table-container--large__info-area__text">$300.00</p>
      </div>
    </div>
  </div>
  <div class="held-history-table-container--small">
    <heldhistoryallstatstableitemsmall-stub heldhistoryitem="[object Object]"></heldhistoryallstatstableitemsmall-stub>
    <heldhistoryallstatstableitemsmall-stub heldhistoryitem="[object Object]"></heldhistoryallstatstableitemsmall-stub>
  </div>
</div>
`;
